### PR TITLE
Add snapshot restore e2e test

### PR DESCRIPTION
**Pull request by ShylajaDevadiga**
#### Proposed Changes ####
Add test for snapshots and restore

#### Verification ####
go test -v -timeout=30m ./snapshotrestore_test.go --serverCount=3 --agentCount=1

#### Linked Issues ####
NA

### CODE REVIEW

File: tests/e2e/snapshotrestore/snapshotrestore_test.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+			cmd := "kubectl get pods --kubeconfig=" + kubeConfigFile
+			res, err := e2e.RunCommand(cmd)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(res).Should(ContainSubstring("test-clusterip"))
+			Expect(res).ShouldNot(ContainSubstring("test-nodeport"))
+		})
+	})
+})
+
+var failed = false
```
Comment (briandowns) [inline tests/e2e/snapshotrestore/snapshotrestore_test.go:177]:
```suggestion
var failed bool
```

**Review by brandond — APPROVED** [review]

File: tests/e2e/snapshotrestore/Vagrantfile [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+NODE_ROLES = (ENV['E2E_NODE_ROLES'] ||
+  ["server-0", "server-1", "server-2", "agent-0", "agent-1"])
+NODE_BOXES = (ENV['E2E_NODE_BOXES'] ||
+  ['generic/ubuntu2004', 'generic/ubuntu2004', 'generic/ubuntu2004', 'generic/ubuntu2004', 'generic/ubuntu2004'])
+GITHUB_BRANCH = (ENV['E2E_GITHUB_BRANCH'] || "master")
+RELEASE_VERSION = (ENV['E2E_RELEASE_VERSION'] || "")
+EXTERNAL_DB = (ENV['E2E_EXTERNAL_DB'] || "etcd")
+HARDENED = (ENV['E2E_HARDENED'] || "")
+REGISTRY = (ENV['E2E_REGISTRY'] || "")
+RANCHER = (ENV['E2E_RANCHER'] || "")
```
Comment (dereknola) [inline tests/e2e/snapshotrestore/Vagrantfile:11]: No need to support any of these Variables for this test, this is an etcd only test that does not deploy any workloads.
So EXTERNAL_DB = "etcd" should be the only option.

File: tests/e2e/snapshotrestore/Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  
+  db_type = getDBType(role, role_num, vm)
+
+  if !HARDENED.empty?
+    vm.provision "Set kernel parameters", type: "shell", path: scripts_location + "/harden.sh"
+    hardened_arg = "protect-kernel-defaults: true\nkube-apiserver-arg: \"enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount\""
+  end 
+  if !REGISTRY.empty?
+    vm.provision "Set private registry", type: "shell", path: scripts_location + "/registry.sh", args: [ "#{NETWORK_PREFIX}.1" ]
+  end
```
Comment (dereknola) [inline tests/e2e/snapshotrestore/Vagrantfile:43]: Remove support for these variables, they are not necessary

File: tests/e2e/snapshotrestore/Vagrantfile [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  if vm.box.to_s.include?("microos")
+    vm.provision 'k3s-reload', type: 'reload', run: 'once'
+    if !EXTERNAL_DB.empty?
+      vm.provision "shell", inline: "docker start #{EXTERNAL_DB}"
+    end
+  end
+  # This step does not run by default and is designed to be called by higher level tools
+  if !RANCHER.empty?
+    vm.provision "Install Rancher", type: "shell", run: "never", path: scripts_location + "/rancher.sh", args: node_ip
+  end
```
Comment (dereknola) [inline tests/e2e/snapshotrestore/Vagrantfile:100]: See above comments

File: tests/e2e/snapshotrestore/Vagrantfile [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      return "cluster-init: true"
+    end
+  elsif ( EXTERNAL_DB == "none" )
+    # Use internal sqlite, only valid for single node clusters
+  else
+    puts "Unknown EXTERNAL_DB: " + EXTERNAL_DB
+    abort
+  end
+  return ""
+end
```
Comment (dereknola) [inline tests/e2e/snapshotrestore/Vagrantfile:115]: This test requires embedded etcd, 
```suggestion
def getDBType(role, role_num, vm)
  if ( EXTERNAL_DB == "" || EXTERNAL_DB == "etcd" )
    if role.include?("server") && role_num == 0
      return "cluster-init: true"
    end
  else
    puts "Unknown EXTERNAL_DB: " + EXTERNAL_DB
    abort
  end
  return ""
end
```

**Review by dereknola — CHANGES_REQUESTED** [review]

**Review by dereknola — APPROVED** [review]

**Comment by dereknola** [comment]
Merging, codespell is marking a false positive

**Pull request merged**